Add Consumer Expenditure Survey and update documentation

baogorek opened the pull request:
* setup.py
    - I added PyYaml and OpenFisca-us so `pytest tests` would run, but I know there's a way to only install those packages when testing (I just didn't get around to figuring it out)
    - Updated the version to 0.1.4 (necessary?) 
* Added scripting example to `README`, added to `@dataset` decorator documentation
* For `external/README.md`, it's just a newline at the end of the file
* Added `ce_raw.py`, `ce.py`, and `tests/ce/test_ce.py`
* Updated all `__init__.py` files and the import tests

**Update**: I reformatted those files with black but still seeing the linter failure. Please let me know if you have any tricks.

MaxGhenis commented on openfisca_us_data/datasets/ce/raw_ce.py:
```diff
@@ -0,0 +1,88 @@
+                    filename = quarter_filenames[quarter - 1]
+                    with zipfile.open(f"{dirstring}/{filename}.csv") as f:
+                        q_df = pd.read_csv(f)
+                        q_df.insert(0, "nominal_year", year)
```
Can't do `q_df["nominal_year"] = year`?

baogorek replied:
Definitely can. It is a wide dataset so I thought it looked better with my chosen set of variables at the front. But, this is just a raw data set to be cleaned up in ce.py.

Your call. If I remove these maybe I should remove the other inserts and let the variables stay where they fall.

MaxGhenis replied:
Yeah I think column reordering for easier visualization could go in a notebook or a separate step.

baogorek replied:
Doing this. Thanks @nikhilwoodruff  for reteaching me the Python 101

MaxGhenis commented on openfisca_us_data/datasets/ce/raw_ce.py:
```diff
@@ -0,0 +1,88 @@
+                            "interview_id",
+                            q_df["NEWID"].apply(lambda x: get_interview_id(x)),
+                        )
+                        q_df.insert(4, "interview_mo", q_df["QINTRVMO"])
```
could these be renames?

baogorek replied:
Yeah they could be. I was just thinking that if someone had the Excel data dictionary in hand, it'd be nice if they could look up every variable in the FMLI dictionary. It's redundant but not a huge space cost. Your call.

MaxGhenis replied:
@nikhilwoodruff wdyt? Is the intent for the raw files to preserve all original columns, or those used for any variables added to the raw file?

Either way I'd suggest a dictionary, which can be used for either a `rename` or a loop to create the new columns.

baogorek replied:
Doing this now

MaxGhenis commented on openfisca_us_data/datasets/ce/raw_ce.py:
```diff
@@ -0,0 +1,88 @@
+                        q_df.insert(
+                            3,
+                            "interview_id",
+                            q_df["NEWID"].apply(lambda x: get_interview_id(x)),
```
```suggestion
                            # Interview ID is the final digit of NEWID.
                            q_df.NEWID.astype(str).str[-1].astype(int)
```

baogorek replied:
Are you thinking of deleting `get_unit_id` and `get_interview_id` and replacing them with one-line comments?

baogorek replied:
Went ahead and did that. Let me know what you think.

MaxGhenis commented on openfisca_us_data/datasets/ce/raw_ce.py:
```diff
@@ -0,0 +1,88 @@
+                        q_df.insert(0, "nominal_year", year)
+                        q_df.insert(1, "nominal_quarter", quarter)
+                        q_df.insert(
+                            2, "cu_id", q_df["NEWID"].apply(lambda x: get_unit_id(x))
```
I think these vectorizations can replace the apply statements (though just as a heads up, I think `q_df.NEWID.apply(get_unit_id)` would also have worked a bit more concisely).
```suggestion
                            2, "cu_id",
                            # CUID is NEWID without the final digit.
                            q_df.NEWID.astype(str).str[:-1].astype(int)
```

MaxGhenis commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+        months_in_scope (int): the number of calendar months (out of a
+            possible 3) that will represent the nominal quarter
+    """
+    months_in_scope = np.nan
```
Don't think we need to initialize since all the cases are either provided or errors right?

baogorek replied:
you're right, we don't.

MaxGhenis commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+            lambda row: months_in_scope(row["interview_mo"], row["nominal_quarter"]),
+            axis=1,
+        )
+        fmli_df.insert(7, "months_in_scope", mis)
```
I see that the 7 follows from other `insert` statements for other column indexes after reviewing to the bottom. Unless the column order is important, I think the standard `df["col"] = vals` would be clearer.

nikhilwoodruff replied:
I agree here

baogorek replied:
done. Also, after realizing that *raw.py uses PyTables, which is good at storing DataFrames in HDF5, I wondered if this logic would be better there. Let me know if you want me to move it.

MaxGhenis commented on tests/ce/test_ce.py:
```diff
@@ -0,0 +1,33 @@
+
+    # Test Household Sector CO2 emissions are "in the ballpark" -----
+    est_co2_kg_per_hh = ce_2019["/annual/co2_kg"][()]
+    hhs_in_us = 122.8e6
```
```suggestion
    HHS_IN_US = 122.8e6
```
same with other constants

baogorek replied:
Made them constants. Also decided to pull them out of the function. Let me know if you think they belong inside

baogorek replied:
Changed my mind and left them in the function

MaxGhenis commented on tests/ce/test_ce.py:
```diff
@@ -0,0 +1,33 @@
+
+    # Test household average demographic estimate -------------------------
+    est_income_bf_tax = ce_2019["/annual/income_before_tax"][()]
+    assert round(est_income_bf_tax) == 82743
```
could you make 82743 a constant and refer to that? same with others here

baogorek replied:
done

MaxGhenis commented on openfisca_us_data/utils.py:
```diff
@@ -22,6 +23,23 @@ def __get__(self, obj, owner):
+    """Decorator function adding functionality to raw data source classes
+
+    Args:
+        cls (a Python class): A python class that has the generate() function
```
```suggestion
        cls (a Python class): A Python class that has the generate() function
```

baogorek replied:
done

MaxGhenis reviewed (commented):
This is awesome, thank you Ben! Appreciate your sprinkles of documentation in the existing code base. I suspect @nikhilwoodruff will have some other ideas but I left a few initial suggestions here. I'll also give it a test run later to see how the data looks.

On black, are you running `black -l 79`? We've set the line length to 79 vs black's default of 88.

nikhilwoodruff commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
 
+### Scripting
+```
+my_acs = ACS()
````
We actually don't need to instantiate the classes - `ACS.generate(2016)` will work, and `ACS` in place of `my_acs` everwhere.

baogorek replied:
fancy! replacing them now...

nikhilwoodruff commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+PosixPath('/mnt/c/devl/openfisca-us-data/openfisca_us_data/microdata/openfisca_us')
+```
+
+If you look inside, there's a auto-generated README file and an `acs_2016.hf file.
````
```suggestion
If you look inside, there's a auto-generated README file and an `acs_2016.h5` file.
```

baogorek replied:
good catch

nikhilwoodruff commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+    """
+    group_prefix = "/household/expenditures/"
+    ce["/household/emissions/co2_kg"] = (
+        1.0 * ce[group_prefix + "airfare"][:]
```
Yeah, or maybe a YAML file would be neater, e.g.
```yaml
airfare: 1.0
alcohol: 0.33
```
```python
with open("yaml_file.yaml") as f:
    categorical_emissions = yaml.load(f)

ce["/household/emissions/c02_kg"] = sum([
    categorical_emissions[category] * ce[group_prefix + category][:]
    for category in categorical_emissions
])
```

nikhilwoodruff replied:
These will eventually be ported to `openfisca-us` too.

baogorek replied:
let me know if I got it close enough. I also had to make modifications to setup.py

nikhilwoodruff commented on setup.py:
```diff
@@ -19,10 +19,12 @@
     ],
     extras_require={
-        "dev": ["autopep8", "black", "setuptools", "wheel", "openfisca-us"]
+        "dev": ["autopep8", "black", "setuptools", "wheel", "openfisca-us"],
```
I think it's fine to move all test-only dependencies into the `dev` requirements (`pyyaml` and `openfisca-us`).

baogorek replied:
Just learned that PyPi is case-insensitive. Do you want me to move "pytest" and "pytest-dependency" into dev or leave in install_requires?

nikhilwoodruff commented on tests/ce/test_ce.py:
```diff
@@ -0,0 +1,33 @@
+
+def test_ce_from_2019():
+    """Test the Consumer Expenditure data generating functions for 2019"""
+    ce = CE()
```
Yeah here we also don't need the instantiation step - everything should work the same from the actual class. It works with the instance of the class too, because all the same attributes are properties.

baogorek replied:
initiation step gone.

nikhilwoodruff reviewed (commented):
Amazing, thanks @baogorek. Left a few more comments. You can also use `make format` as a shortcut to our combination of `autopep8` (for the comment formatting) and `black`.

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
 ```
 
+### Scripting
+```
````
done

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+
+After successful running of the command above, the data has been stored. The `data_dir` property
+shows where:
+```
````
done

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+shows where:
+```
+my_acs.data_dir
+PosixPath('/mnt/c/devl/openfisca-us-data/openfisca_us_data/microdata/openfisca_us')
````
done

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+Note that it's 196 MB, so it contains some data. We can load that data (still in HDF5 format)
+with the `load()` method.
+
+```
````
done

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+### Scripting
+```
+my_acs = ACS()
+my_acs.generate(2016)  # retrieves the data
````
done

baogorek commented:
Thanks @MaxGhenis and @nikhilwoodruff ! I resolved most things, left a few follow-up questions, and added a bit more to the README. Saving the YAML coefficients suggestion for tomorrow morning.

baogorek commented on README.md:
````diff
@@ -25,6 +29,47 @@ For example (doesn't work yet):
+```
+my_acs = ACS()
+acs_hd5 = my_acs.load(2016)
+
````
done

baogorek commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+        """Saves organized Consumer Expenditure data in HDF5 format via h5py
+
+        Computes "months in scope" from existing variables, and uses it to
+        adjust weights and compute annual estimates. Finally, those estimates
```
added them.

baogorek commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+
+        raw_data = RawCE.load(year)
+        ce = h5py.File(CE.file(year), mode="w")
+        ce.create_group("/household")  # Household quarterly data
```
Ok

baogorek commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,279 @@
+    Args:
+        ce (h5py.File): The HDF5 data structure containing the organized
+           Consumer Expenditure survey data
+        fmli_df (pd.DataFrame): The raw CE data (all 5 quarters)
```
it's not

baogorek commented:
Got the YAML file working (required edits to setup.py). I think I've addressed all comments (some questions remaining). Let me know if I forgot something.

baogorek commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,280 @@
+class CE:
+    """OpenFisca-US dataset for the Consumer Expenditure Survey
+
+    The information extracted by RawCE and CE classes is from the "Interview
```
@MaxGhenis , somewhere earlier you had asked about documenting use of the FMLI files. Let me know if this is sufficient.

MaxGhenis replied:
thanks, could you also spell out what FMLI abbreviates?

baogorek replied:
Believe it or not, I don't think it's an abbreviation for anything. From [here](https://www.ilo.org/surveyLib/index.php/catalog/350/download/3928), 

> The FMLI file, also referred to as the “Consumer Unit Characteristics and Income” file, contains
> CU characteristics, CU income, and characteristics and earnings of the reference person and of the
> spouse. The file includes weights needed to calculate population estimates and variances.

Let me know if you can find something on the name, or if you have other suggestions about how to describe it.

MaxGhenis replied:
Weird! OK thanks.

MaxGhenis commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,280 @@
+    #     pretax income, Consumer Expenditure Surveys, 2016.
+    MEAN_CASH_CONTRIB_OUTSIDE_HOUSEHOLD_2016 = 2080.85
+    MEAN_CASH_CONTRIB_TO_CHARITIES_2016 = 396.59
+    PROPORTION_CASH_CONTRIB_TO_CHARITY = (
```
this can be lowercase

MaxGhenis commented on openfisca_us_data/datasets/ce/ce.py:
```diff
@@ -0,0 +1,280 @@
+    quarterly household level (technically "consumer unit") data deemed most
+    appropriate for this project's needs. Only one set of files from the
+    Interview Survey, called "FMLI" is used, and contains quarterly
+    expenditures, annual income, assets, and liablities, and household
```
```suggestion
    expenditures, annual income, assets, and liabilities, and household
```

baogorek replied:
fixed

MaxGhenis commented on tests/ce/test_ce.py:
```diff
@@ -0,0 +1,46 @@
+    limitations. These were considered sufficiently close for the purposes
+    of simulation.
+    """
+    # 2019 Constants -------------------------------------------------------------
```
```suggestion
    # 2019 constants, calculated from the CE -------------------------------------------------------------
```

baogorek replied:
done

MaxGhenis commented on tests/ce/test_ce.py:
```diff
@@ -0,0 +1,46 @@
+    of simulation.
+    """
+    # 2019 Constants -------------------------------------------------------------
+    HHS_IN_US = 122.8e6
```
Is this also from the CE? If so would it be CUs?

baogorek replied:
I guess it's from out of the blue. I googled number of households in the US, just to demonstrate that we're in the ballpark for annual CO2. I can remove it, rename it, ..., your call.

MaxGhenis replied:
This works, could you just add the source?

baogorek replied:
left some more documentation. You can see in the calculations that we're multiplying a per_cu quantity with number of households, which is suggestive at least

baogorek commented:
Haven't forgotten about this. A few things came up. Will get back to it soon

baogorek commented:
@nikhilwoodruff , @MaxGhenis , would you help me out with this - what am I doing wrong?

Relates to lines 66 to 73 in raw_ce.py, just can't seem to assign cu_id and interview_id
```
import pandas as pd
my_df = pd.DataFrame({'x': ["124", "5325"], 'y': [123, 535]})
r = my_df.x.astype(str).str[:-1].astype(int)
my_df[["r"]] = r
```
leads to 
```
ValueError: Columns must be same length as key
```

nikhilwoodruff commented:
@baogorek not sure if I'm missing something but could it be:
```python
my_df["r"] = r
```
instead of 
```python
my_df[["r"]] = r
```

baogorek commented:
Just checking in. Anything else that I should do?

nikhilwoodruff reviewed (approved):
Just re-checked this- I think it's good to merge! As mentioned, we'll also add the carbon logic and parameters eventually to OpenFisca-US (like in the UK model), plus also making sure all the variables here are defined in the model. But I think this PR is complete at its objective. @MaxGhenis WDYT?

MaxGhenis commented:
Yep, looks great. Thank you @baogorek!
